packages/eslint-plugin/src/rules/unbound-method.ts

```diff
@@ -209,6 +214,42 @@ export default createRule<Options, MessageIds>({
               );
             }
           });
+          return;
```

Comment:
The control flow here has a couple gotchas surrounding 
1. variables with type annotations
2. declarations (via `declare` modifier or `FunctionDeclaration`)

```ts
class Foo {
  unbound = function () { };
}

// not sure why this is even legal syntax tbh, but this is a regression (shouldn't flag):
declare const { unboundMethod }: Foo; // ditto for var/let
                ^^^^^^^^^^^^^

// also should not flag
declare const { split } = "wtf";
                ^^^^^

interface HasMethod {
   method(): void;
}

// bug with current impl; rule optimistically uses the `initNode` but should still check the type annotation.
function shouldFlag({ method }: HasMethod = { method: () => {} }) {};
                      ^^^^^^

// declarations strike again. (this is just a signature, not runtime code)
function shouldNotFlag({ method }: HasMethod);
                         ^^^^^^

// Should this flag? I don't think so... but I could be convinced either way. 
const { method }: HasMethod = { method: () => {} };
```

Reply:
A lot of nice catches, thanks!